src/VisualStudio/Core/Def/Implementation/SymbolSearch/SymbolSearchSource.cs

```diff
@@ -0,0 +1,99 @@
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.SymbolSearch
+{
+    internal class SymbolSearchSource : ISymbolSourceFromLocation
+    {
+        internal SymbolSearchSourceProvider ServiceProvider { get; }
+
+        public SymbolSearchSource(SymbolSearchSourceProvider symbolSourceProvider)
+        {
+            ServiceProvider = symbolSourceProvider;
+        }
+
+        public ImageId Icon => default;
+
+        public string DescriptionText => string.Empty;
+
+        string INamed.DisplayName => ServicesVSResources.Symbol_search_source_name;
+
+        string ISymbolSource.Id => nameof(SymbolSearchSource);
+
+        public async Task<SymbolSearchStatus> FindSymbolsAsync(string navigationKind, VisualStudio.Language.Intellisense.SymbolSearch.Location sourceLocation, ISymbolSearchCallback callback, CancellationToken token)
+        {
+            // This method is called off the UI thread
+            var snapshot = sourceLocation.PersistentSpan.Document.TextBuffer.CurrentSnapshot;
+            var roslynDocument = snapshot.GetOpenDocumentInCurrentContextWithChanges();
+
+            var solutionPath = roslynDocument.Project.Solution.FilePath;
+            var rootNodeName = !string.IsNullOrWhiteSpace(solutionPath)
+                ? string.Format(ServicesVSResources.Symbol_search_known_solution, Path.GetFileNameWithoutExtension(solutionPath))
+                : ServicesVSResources.Symbol_search_current_solution;
+
+            var symbolSearchContext = new SymbolSearchContext(this, callback, rootNodeName, token);
+
+            try
+            {
+                switch (navigationKind)
+                {
+                    case PredefinedNavigationKinds.Definition:
+                        {
+                            // This is not yet supported,
+                            // because Roslyn implementation depends on UI thread.
+                            return SymbolSearchStatus.Withheld;
+                            // Existing Roslyn implementation, disabled to not interfere with demos
+                            /*
+                            var goToDefinitionService = roslynDocument.GetLanguageService<IGoToDefinitionService>();
+                            goToDefinitionService.TryGoToDefinition(roslynDocument, sourceLocation.PersistentSpan.Span.GetStartPoint(snapshot).Position, token);
+                            return SymbolSearchStatus.Completed;
+                            */
+                        }
+                    case PredefinedNavigationKinds.Implementation:
+                        {
+                            var findUsagesService = roslynDocument.GetLanguageService<IFindUsagesService>();
+                            var context = new SimpleFindUsagesContext(token);
+                            await findUsagesService.FindImplementationsAsync(roslynDocument, sourceLocation.PersistentSpan.Span.GetStartPoint(snapshot).Position, symbolSearchContext)
```

Comment:
Extract out the sourceLocation.PersistentSpan.Span.GetStartPoint(snapshot).Position bit?